source/opt/workaround_image_operands.cpp

```diff
@@ -0,0 +1,173 @@
+            // Specifically we care about OpTypeSampledImage calls
+            analysis::SampledImage* sampled_image_type =
+                type_mgr->GetType(inst->result_id())->AsSampledImage();
+
+            // Find all of the uses of these OpTypeSampledImages
+            def_use_mgr->ForEachUse(
+                inst, [this, &fixed_instructions, def_use_mgr, img_type,
+                       type_mgr, sampled_image_type,
+                       &modified](ir::Instruction* op_type_sampled_image_usage,
+                                  uint32_t idx) {
+                  if (op_type_sampled_image_usage->opcode() ==
+                      SpvOpSampledImage) {
+                    // Specifically OpSampledImage values.
+                    auto op_sampled_image = op_type_sampled_image_usage;
+                    def_use_mgr->ForEachUse(
+                        op_sampled_image,
+                        [this, &fixed_instructions, op_sampled_image,
+                         sampled_image_type, def_use_mgr, img_type, type_mgr,
+                         &modified](ir::Instruction* sampled_image_usage,
+                                    uint32_t idx) {
+                          bool has_depth = false;
+                          if (sampled_image_usage->opcode() ==
+                                  SpvOpImageDrefGather ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSampleDrefImplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSampleDrefExplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSampleProjDrefImplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSampleProjDrefExplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSparseSampleDrefImplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSparseSampleDrefExplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSparseSampleProjDrefImplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSparseSampleProjDrefExplicitLod ||
+                              sampled_image_usage->opcode() ==
+                                  SpvOpImageSparseDrefGather) {
+                            has_depth = true;
+                          }
+                          // If any of these are a dref lookup
+                          if (!has_depth) {
+                            return;
+                          }
+                          analysis::Type* t = type_mgr->GetType(
+                              type_mgr->GetId(img_type->sampled_type()));
```

Comment:
Why not set `t` to `img_type->sampled_type()`?  Is there a reason `t` is not const?

Reply:
The analysis::Image constructor takes a non-const analysis::Type*. 
This may have just been accidental (I doesn't LOOK like it needs a mutable type), but I was avoiding making external changes. I can make Image take a only hold onto a const Type, which would simplify this.

Reply:
There is some inconsistency in the use of const in the type manager.  This is fine for now.  We should have a follow up issue to fix it.  I have a PR #1584 that does some fix up.